File: src/simulation/GBSensorShot.java
/*******************************************************************************
 * Copyright (c) 2002-2013 (c) Devon and Warren Schudy
 * Copyright (c) 2014  Devon and Warren Schudy, Mike Anderson
 *******************************************************************************/
package simulation;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Rectangle;

import Rendering.GBProjection;
import sides.Side;
import support.FinePoint;

class GBSensorShot extends GBObject {
	GBRobot owner;
	Side side;
	GBSensorState state;
	GBObjectClass seen;
	int age;
	FinePoint focus;

	// public:

	public static final int kSensorShotMinLifetime = 15;

	int getLifetime() {
		return (int) Math.floor(Math.max(radius, kSensorShotMinLifetime));
	}

	public GBSensorShot(FinePoint fcs, GBRobot who, GBSensorState st) {
		super(who.getPosition(), st.getMaxRange());
		owner = who;
		side = who.getOwner();
		state = st;
		seen = st.getSeen();
		focus = fcs;
	}

	@Override
	public void collideWith(GBObject other) {
		if (other.getObjectClass() == GBObjectClass.ocRobot && (GBRobot) other == owner)
			return; // Seeing self is never allowed
		state.report(new GBSensorResult(other, (other.getPosition()
				.subtract(focus)).norm())); // most logic is now in SensorState
	}

	@Override
	public void act(GBWorld world) {
		age++;
	}

	@Override
	public GBObjectClass getObjectClass() {
		if (age >= getLifetime())
			return GBObjectClass.ocDead;
		else
			return GBObjectClass.ocSensorShot;
	}

	@Override
	public String toString() {
		String classname;
		switch (seen) {
		case ocRobot:
			classname = "Robot";
			break;
		case ocFood:
			classname = "Food";
			break;
		case ocShot:
			classname = "Shot";
			break;
		default:
			classname = "Mystery";
			break;
		}
		return classname + " sensor for " + owner.toString();
	}

	@Override
	public Color getColor() {
		float fraction = 1.0f - age / getLifetime();
		switch (seen) {
		case ocRobot:
			return new Color(0.4f * fraction, 0.8f * fraction, fraction);
		case ocFood:
			return new Color(0.5f * fraction, fraction, 0.5f * fraction);
		case ocShot:
			return new Color(fraction, fraction, 0.5f * fraction);
		default:
			return new Color(fraction, fraction, fraction);
		}
	}

	public GBRobot getFirer() {
		return owner;
	}

	public GBObjectClass getSeen() {
		return age == 0 ? seen : GBObjectClass.ocDead;
	}

	@Override
	public Side getOwner() {
		return side;
	}

	@Override
	public void draw(Graphics g, GBProjection proj, boolean detailed) {
		// show focus, owner, and side?
		Graphics2D g2d = (Graphics2D) g;
		Rectangle where = getScreenRect(proj);
		g2d.setColor(getColor());
		g2d.setStroke(new BasicStroke(1));
		g2d.drawOval(where.x, where.y, where.width, where.height);
	}
}

File: grobots_cpp/simulation/GBSensorShot.cpp
// GBSensorShot.cpp
// Grobots (c) 2002-2004 Devon and Warren Schudy
// Distributed under the GNU General Public License.

#include "GBSensorShot.h"
#include "GBHardwareState.h"
#include "GBRobot.h"

const GBFrames kSensorShotMinLifetime = 15;


GBFrames GBSensorShot::Lifetime() const {
	return floor(max(radius, kSensorShotMinLifetime));
}

GBSensorShot::GBSensorShot(const GBPosition & fcs, GBRobot * who, GBSensorState * st)
	: GBObject(who->Position(), st->MaxRange()),
	owner(who), side(who->Owner()),
	state(st),
	seen(st->Seen()),
	age(0),
	focus(fcs)
{}

void GBSensorShot::CollideWith(GBObject * other) {
	if ( other->Class() == ocRobot && (GBRobot *)other == owner )
		return; //Seeing self is never allowed
	state->Report(GBSensorResult(other, (other->Position() - focus).Norm())); //most logic is now in SensorState
}

void GBSensorShot::Act(GBWorld *) {
	age ++;
}

GBObjectClass GBSensorShot::Class() const {
	if ( age >= Lifetime() )
		return ocDead;
	else
		return ocSensorShot;
}

string GBSensorShot::Description() const {
	string classname;
	switch ( seen ) {
		case ocRobot: classname = "Robot"; break;
		case ocFood: classname = "Food"; break;
		case ocShot: classname = "Shot"; break;
		default: classname = "Mystery"; break;
	}
	return classname + " sensor for " + owner->Description();
}

const GBColor GBSensorShot::Color() const {
	float fraction = 1.0 - (float)age / Lifetime();
	switch ( seen ) {
		case ocRobot: return GBColor(0.4f, 0.8f, 1) * fraction;
		case ocFood: return GBColor(0.5f, 1, 0.5f) * fraction;
		case ocShot: return GBColor(1, 1, 0.5f) * fraction;
	}
	return GBColor(fraction);
}

const GBRobot * GBSensorShot::Firer() const {
	return owner;
}

const GBObjectClass GBSensorShot::Seen() const {
	return age == 0 ? seen : ocDead;
}

GBSide * GBSensorShot::Owner() const { return side; }

void GBSensorShot::Draw(GBGraphics & g, const GBRect & where, bool /*detailed*/) const {
	// show focus, owner, and side?
	g.DrawOpenOval(where, Color());
}

void GBSensorShot::DrawMini(GBGraphics & g, const GBRect & where) const {
	g.DrawOpenOval(where, Color());
}

